odometry.c

```diff
@@ -0,0 +1,190 @@
+							 * cos(robot->pose.theta + 0.5f * delta_rot) / dt;
+		robot->velocity.y = delta_fwd
+							 * sin(robot->pose.theta + 0.5f * delta_rot) / dt;
+		robot->velocity.omega = delta_rot / dt;
+	}
+
+	robot->pose.x += delta_fwd * cos(robot->pose.theta + 0.5f * delta_rot);
+	robot->pose.y += delta_fwd * sin(robot->pose.theta + 0.5f * delta_rot);
+	robot->pose.theta += delta_rot;
+
+	robot->time_last_estim = time_now;
+}
+
+void wheel_init(
+		odometry_wheel_t *wheel,
+		const float wheel_radius)
+{
+	wheel->tick_to_meter = (2 * M_PI * wheel_radius / 65536);
+	wheel->delta_pos_accumulator = 0;
+	wheel->is_initialised = false;
+}
+
+void wheel_update(
+		odometry_wheel_t *wheel,
+		const odometry_encoder_sample_t new_sample)
+{
+	// If not initialised, Initialise!
+	if (!(wheel->is_initialised)) {
+		wheel->samples[0].timestamp = new_sample.timestamp;
+		wheel->samples[0].value = new_sample.value;
+		wheel->samples[1].timestamp = new_sample.timestamp;
+		wheel->samples[1].value = new_sample.value;
+		wheel->samples[2].timestamp = new_sample.timestamp;
+		wheel->samples[2].value = new_sample.value;
+
+		wheel->delta_pos_accumulator = 0;
+		wheel->is_initialised = true;
+	} else {
+		wheel->samples[0].timestamp = wheel->samples[1].timestamp;
+		wheel->samples[0].value = wheel->samples[1].value;
+		wheel->samples[1].timestamp = wheel->samples[2].timestamp;
+		wheel->samples[1].value = wheel->samples[2].value;
+		wheel->samples[2].timestamp = new_sample.timestamp;
+		wheel->samples[2].value = new_sample.value;
+
+		wheel->delta_pos_accumulator += (int16_t) (wheel->samples[2].value
+								    			  - wheel->samples[1].value);
+	}
+}
+
+uint16_t wheel_predict(
+		odometry_wheel_t *wheel,
+		const timestamp_t time_now)
+{
+	if (wheel->samples[2].timestamp != time_now) {
+		float acc, vel, pos, tau1, tau2, dt1, dt2, dt;
+		int16_t dy1, dy2;
+		uint16_t prediction;
+
+		// Fit a parabola to the wheel's motion
+		dt1 = timestamp_duration_s(wheel->samples[0].timestamp,
+								   wheel->samples[1].timestamp);
+		dt2 = timestamp_duration_s(wheel->samples[0].timestamp,
+								   wheel->samples[2].timestamp);
+		dy1 = (int16_t) (wheel->samples[1].value - wheel->samples[0].value);
+		dy2 = (int16_t) (wheel->samples[2].value - wheel->samples[0].value);
+		tau1 = dy1 / (- dt1 * dt1 + dt1 * dt2);
+		tau2 = dy2 / (- dt2 * dt2 + dt1 * dt2);
+
+		acc = - tau1 - tau2;
+		vel = tau1 * dt2 + tau2 * dt1;
+		pos = wheel->samples[0].value;
+
+		// Predict change from last encoder value recorded
+		dt = timestamp_duration_s(wheel->samples[0].timestamp, time_now);
+		prediction = pos + (int16_t) (vel * dt + acc * dt * dt);
+
+		return prediction;
+	} else {
+		return wheel->samples[2].value;
+	}
+}
+
+int16_t wheel_get_delta_tick(
+		odometry_wheel_t *wheel,
+		const timestamp_t time_now)
+{
+	int16_t delta;
+	uint16_t prediction;
+	prediction = wheel_predict(wheel, time_now) - wheel->samples[2].value;
+	delta = wheel->delta_pos_accumulator + prediction;
```

Comment:
```delta = prediction - accumulator```, and it would be more logical if accumulator were uint16_t. (Although I think it should would work with int16_t since it is converted to uint16_t by the type promotion rules)